crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_gpu.rs

```diff
@@ -0,0 +1,407 @@
+/// and scalars required to call Blitzar's `vlen_msm` function.
+#[tracing::instrument(name = "create_blitzar_metadata_tables", level = "debug", skip_all)]
+fn create_blitzar_metadata_tables(
+    committable_columns: &[CommittableColumn],
+    offset: usize,
+) -> (Vec<u32>, Vec<u32>, Vec<u8>) {
+    // Keep track of the lengths of the columns to handled signed data columns.
+    let ones_columns_lengths = committable_columns
+        .iter()
+        .map(CommittableColumn::len)
+        .collect_vec();
+
+    // The maximum matrix size will be used to create the scalars vector.
+    let (max_height, max_width) = if let Some(max_column_len) =
+        committable_columns.iter().map(CommittableColumn::len).max()
+    {
+        matrix_size(max_column_len, offset)
+    } else {
+        (0, 0)
+    };
+
+    // Find the single packed byte size of all committable columns.
+    let num_of_bytes_in_committable_columns: usize = committable_columns
+        .iter()
+        .map(|column| column.column_type().byte_size())
+        .sum();
+
+    // Get a single bit table entry with ones added for all committable columns that are signed.
+    let single_entry_in_blitzar_output_bit_table: Vec<u32> = committable_columns
+        .iter()
+        .map(|column| column.column_type().bit_size())
+        .chain(iter::repeat(BYTE_SIZE).take(ones_columns_lengths.len()))
+        .collect();
+
+    // Create the full bit table vector to be used by Blitzar's vlen_msm algorithm.
+    let blitzar_output_bit_table: Vec<u32> = single_entry_in_blitzar_output_bit_table
+        .iter()
+        .copied()
+        .cycle()
+        .take(single_entry_in_blitzar_output_bit_table.len() * max_height)
+        .collect();
+
+    // Create the full length vector to be used by Blitzar's vlen_msm algorithm.
+    let blitzar_output_length_table: Vec<u32> = (0..blitzar_output_bit_table.len()
+        / single_entry_in_blitzar_output_bit_table.len())
+        .flat_map(|i| {
+            itertools::repeat_n(
+                full_width_of_row(i) as u32,
+                single_entry_in_blitzar_output_bit_table.len(),
+            )
+        })
+        .collect();
+
+    // Create a cumulative length table to be used when packing the scalar vector.
+    let cumulative_byte_length_table: Vec<usize> = iter::once(0)
+        .chain(blitzar_output_bit_table.iter().scan(0usize, |acc, &x| {
+            *acc += (x / BYTE_SIZE) as usize;
+            Some(*acc)
+        }))
+        .collect();
+
+    // Create scalars array. Note, scalars need to be stored in a column-major order.
+    let num_scalar_rows = max_width;
+    let num_scalar_columns =
+        (num_of_bytes_in_committable_columns + ones_columns_lengths.len()) * max_height;
+    let mut blitzar_scalars = vec![0u8; num_scalar_rows * num_scalar_columns];
+
+    // Populate the scalars array.
+    let span = span!(Level::INFO, "pack_blitzar_scalars").entered();
+    if !blitzar_scalars.is_empty() {
```

Comment:
NIT: this `if` doesn't do anything. The chunks_exact_mut will be an empty iterator if it is empty.

Reply:
Never mind, this doesn't work.